File: java/array/contains_duplicate_2.java
// 219. Contains Duplicate II
// https://leetcode.com/problems/contains-duplicate-ii/description/

import java.util.*;

class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        // Create hashset to store unique elements from the array
        HashSet<Integer> set = new HashSet<>();

        // iterate through each element in array
        for (int i=0;i<nums.length;i++){
            // check if the element is already in the hashset
            if(set.contains(nums[i])){
                return true;
            }
            set.add(nums[i]);

            // when lenght > k remove first element form hashset
            if (set.size() > k){
                set.remove(nums[i-k]);
            }
        }
        return false; // no duplicate
    }
}






public class contains_duplicate_2 {
    public static void main (String[] args) {
        int[] nums = {1,2,3,1};
        int k = 3;
        Solution sol = new Solution();
        System.out.println(sol.containsNearbyDuplicate(nums,k));
    }
}

File: python/array/contains_duplicate_2.py
# 219. Contains Duplicate II
# https://leetcode.com/problems/contains-duplicate-ii/description/

from typing import List, Dict, Tuple, Optional

class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        # Create a set to store unique elements in the sliding window
        seen_nums = set()
        # Iterate through the array
        for i in range(len(nums)):
            # Check if current number already exists in set
            if nums[i] in seen_nums:
                return True

            # Add current number to set
            seen_nums.add(nums[i])

            # Maintain the sliding window of size k
            if len(seen_nums) > k:
                seen_nums.remove(nums[i-k])
        return False # No duplicates found within distance k






if __name__ == '__main__':
    nums = [1, 2, 3, 1]
    k = 3
    sol = Solution()
    print(sol.containsNearbyDuplicate(nums, k))
